### PR TITLE
planner: EXPLAIN will make SET_VAR not SQL level

### LINKED ISSUE #46888: SET_VAR hint not work in global binding
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Add following test in `bindinfo/tests/bind_test.go`.

```go
func TestStmtHints2(t *testing.T) {
	store := testkit.CreateMockStore(t)

	tk := testkit.NewTestKit(t, store)
	tk.MustExec("use test")
	tk.MustExec("drop table if exists t")
	tk.MustExec("create table t(a int, b int, index idx(a))")
	tk.MustQuery("select @@MAX_EXECUTION_TIME").Check(testkit.Rows("0"))
	tk.MustExec("create global binding for select * from t using select /*+ set_var(max_execution_time=100), MEMORY_QUOTA(2 GB) */ * from t use index(idx)")
	tk.MustQuery("select * from t")
	require.Equal(t, int64(2147483648), tk.Session().GetSessionVars().MemTracker.GetBytesLimit())
	require.Equal(t, uint64(100), tk.Session().GetSessionVars().StmtCtx.MaxExecutionTime)
	tk.MustQuery("select a, b from t")
	require.Equal(t, int64(1073741824), tk.Session().GetSessionVars().MemTracker.GetBytesLimit())
	require.Equal(t, uint64(0), tk.Session().GetSessionVars().StmtCtx.MaxExecutionTime)
}
```

### 4. What is your TiDB version? (Required)

master: 22f1286095b53a9d8e2eeacc2ea024d449dc6351

### LINKED ISSUE #46958: EXPLAIN ANALYZE doesn't restore the variable set by `SET_VAR`
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
mysql> select @@tidb_executor_concurrency;
+-----------------------------+
| @@tidb_executor_concurrency |
+-----------------------------+
|                           5 |
+-----------------------------+
1 row in set (0.00 sec)

mysql> select /*+ set_var(tidb_executor_concurrency=8) */ @@tidb_executor_concurrency;
+-----------------------------+
| @@tidb_executor_concurrency |
+-----------------------------+
|                           8 |
+-----------------------------+
1 row in set (0.00 sec)

mysql> select @@tidb_executor_concurrency;
+-----------------------------+
| @@tidb_executor_concurrency |
+-----------------------------+
|                           5 |
+-----------------------------+
1 row in set (0.00 sec)

mysql> explain analyze select /*+ set_var(tidb_executor_concurrency=8) */ @@tidb_executor_concurrency;
+-------------------+---------+---------+------+---------------+-----------------------------------------------------+---------------+---------+------+
| id                | estRows | actRows | task | access object | execution info                                      | operator info | memory  | disk |
+-------------------+---------+---------+------+---------------+-----------------------------------------------------+---------------+---------+------+
| Projection_3      | 1.00    | 1       | root |               | time:6.18µs, loops:2, RU:0.000000, Concurrency:OFF  | 8->Column#1   | 0 Bytes | N/A  |
| └─TableDual_4     | 1.00    | 1       | root |               | time:1.1µs, loops:2                                 | rows:1        | N/A     | N/A  |
+-------------------+---------+---------+------+---------------+-----------------------------------------------------+---------------+---------+------+
2 rows in set (0.00 sec)

mysql> select @@tidb_executor_concurrency;
+-----------------------------+
| @@tidb_executor_concurrency |
+-----------------------------+
|                           8 |
+-----------------------------+
1 row in set (0.00 sec)
```

### 2. What did you expect to see? (Required)

The final output should be still be the default value.

### 3. What did you see instead (Required)

It's value set by the SET_VAR

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #46958, close #46888

Problem Summary:

### What is changed and how it works?

The explain stmt will enter the Optimize twice. Make the func `AddSetVarHintRestore` add the already changed value into the restore map.
image

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by crazycs520 — APPROVED** [review]

File: executor/bind.go [hunk 1]
```diff
@@ -136,6 +136,10 @@ func (e *SQLBindExec) createSQLBind() error {
 	// is necessary to avoid 'create binding' been recorded as 'explain'.
 	saveStmtCtx := e.Ctx().GetSessionVars().StmtCtx
 	defer func() {
+		// But we need to restore the SET_VAR's setting.
+		for name, val := range e.Ctx().GetSessionVars().StmtCtx.SetVarHintRestore {
+			saveStmtCtx.AddSetVarHintRestore(name, val)
+		}
```
Comment (time-and-fate) [inline executor/bind.go:142]: I suggest we add some comments on why this is needed, like where is the `SetVarHintRestore` set.
It's really not easy to realize.

Comment (time-and-fate) [reply executor/bind.go:142]: And looks like we are solving two problems in this PR. I think it's better to clarify it in the issue or PR description.

**Review by time-and-fate — APPROVED** [review]
Others LGTM.

**Review by yudongusa — APPROVED** [review]
LGTM for variables part

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**